### PR TITLE
Implements setting number of nodes for control-plane and workers

### LINKED ISSUE #3145: Enhance cluster node count capability when creating clusters
With the introduction of:

* https://github.com/vmware-tanzu/community-edition/pull/3139

Users can now specify multi-node clusters. However, the current model operates under an assumption that you're going to specify a "number of nodes". Which may be too generic/limiting for future use cases and especially for future providers.

While a little more complex, I think we need to update the API to be:

* `workerNodeCount:`
* `controlPlaneNodeCount:`

It is up to the cluster provider(s) to understand and operate on these, but we should be passing them down as separate values.

One of the primary complexities here is how to express single node clusters. Single node clusters mean there is no distinction between a worker and controlplane node. On a technical level, it's just a control plane node with its taint removed, which would normally prevent scheduling workloads.

Open to ideas on how to express this. For one, single node should be the **default**. And we can make some contractual conventions such as `controlPlaneNodeCount: 1` and `workerNodeCount: 0` is assumed to be a single node cluster. However, this, again, should be a convention implemented in the provider.

### PR DESCRIPTION
**Pull request by jpmcb**
## What this PR does / why we need it

This commit enables `unmanaged-clusters` to support _both_
setting control-plane nodes _and_ worker nodes, not just an
arbitrary number of nodes (where 1 automatically
is set to a control plane)

## Details for the Release Notes (PLEASE PROVIDE)

```release-note
unmanaged-cluster can set both/either control-plane nodes and worker nodes
```

## Which issue(s) this PR fixes
Fixes: #3145

## Describe testing done for PR
```shell
# default behavior preserved ...
$ tanzu unmanaged-cluster create test1
...

$ kubectl get nodes
NAME                  STATUS   ROLES                  AGE     VERSION
test1-control-plane   Ready    control-plane,master   48s     v1.21.5

$ tanzu unmanaged-cluster delete test1
...

# can set any number of control plane and worker nodes
$ tanzu unmanaged-cluster create life --control-plane-node-count 2 --worker-node-count 2
...

$ kubectl get pods -A
NAMESPACE            NAME                                          READY   STATUS    RESTARTS   AGE
kube-system          antrea-agent-ft87c                            2/2     Running   0          3m40s
kube-system          antrea-agent-gpsbd                            2/2     Running   0          3m40s
kube-system          antrea-agent-qks7d                            2/2     Running   0          3m40s
kube-system          antrea-agent-szp6x                            2/2     Running   0          3m40s
kube-system          antrea-controller-67846968bb-5b5zq            1/1     Running   0          3m40s
kube-system          coredns-558bd4d5db-cgn4q                      1/1     Running   0          6m32s
kube-system          coredns-558bd4d5db-kwzg4                      1/1     Running   0          6m32s
kube-system          etcd-life-control-plane                       1/1     Running   0          6m31s
kube-system          etcd-life-control-plane2                      1/1     Running   0          6m16s
kube-system          kube-apiserver-life-control-plane             1/1     Running   0          6m31s
kube-system          kube-apiserver-life-control-plane2            1/1     Running   0          6m17s
kube-system          kube-controller-manager-life-control-plane    1/1     Running   1          6m31s
kube-system          kube-controller-manager-life-control-plane2   1/1     Running   0          6m18s
kube-system          kube-proxy-87wmg                              1/1     Running   0          5m25s
kube-system          kube-proxy-c5zbg                              1/1     Running   0          5m25s
kube-system          kube-proxy-dv5s7                              1/1     Running   0          6m33s
kube-system          kube-proxy-vpp4g                              1/1     Running   0          6m18s
kube-system          kube-scheduler-life-control-plane             1/1     Running   1          6m31s
kube-system          kube-scheduler-life-control-plane2            1/1     Running   0          6m17s
local-path-storage   local-path-provisioner-547f784dff-2qvs2       1/1     Running   0          6m32s
tkg-system           kapp-controller-f7dbbcb4f-szzlf               1/1     Running   0          5m4s


$ kubectl get nodes
NAME                  STATUS   ROLES                  AGE     VERSION
life-control-plane    Ready    control-plane,master   6m48s   v1.21.5
life-control-plane2   Ready    control-plane,master   6m22s   v1.21.5
life-worker           Ready    <none>                 5m29s   v1.21.5
life-worker2          Ready    <none>                 5m29s   v1.21.5

```

---

Note: Per Sean's comments, this does _not_ support users deploying multiple control planes with no worker nodes. It will error out. After this merges, I will create an issue to support removing the taint from the control plane nodes if there are no worker nodes present.

### CODE REVIEW

File: cli/cmd/plugin/unmanaged-cluster/cmd/create.go
Comment (stmcginnis) [inline cli/cmd/plugin/unmanaged-cluster/cmd/create.go:67]: Should these be `StringVar` instead of `StringVarP`? Or still determining what to use for the one-letter shorthand?

File: cli/cmd/plugin/unmanaged-cluster/cluster/kind.go
Comment (stmcginnis) [inline cli/cmd/plugin/unmanaged-cluster/cluster/kind.go:159]: grammar nit: s/nodes/node/

File: cli/cmd/plugin/unmanaged-cluster/config/config_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/unmanaged-cluster/config/config_test.go:222]: Error string on both of these looks like it was copied from the CNI assertion?

**Review by stmcginnis — COMMENTED** [review]
I'm getting a hang if I do `unmanaged-cluster create --control-plane-nodes 2 foo` when it gets to installing kapp-controller. Not sure how that would be related to this though.

We don't have full docs yet for `unamanged-cluster` but one thing we may want to note is that if there are more than 1 control plane nodes requested, there will also be a haproxy load balancer container deployed.

Other than the kapp-controller thing, I've built locally and confirmed multiple nodes are created for worker and control plane node requests.

**Review by joshrosso — CHANGES_REQUESTED** [review]
just adding myself as a reviewer.

Thanks for knocking this out!

**Comment by stmcginnis** [comment]
OK, a little more digging on the kapp-controller hang. I see this:

```
Events:
  Type     Reason            Age    From               Message
  ----     ------            ----   ----               -------
  Warning  FailedScheduling  9m49s  default-scheduler  0/2 nodes are available: 2 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
  Warning  FailedScheduling  9m48s  default-scheduler  0/2 nodes are available: 2 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
```

Looks like we may need to add some more logic around things where if there are no worker nodes requested, but more than one control plane node, we tell it not to taint the nodes.

From single node:

```
$ k get node --show-labels
NAME                STATUS   ROLES                  AGE     VERSION
moo-control-plane   Ready    control-plane,master   2m51s   v1.21.5
~
$ k describe node moo-control-plane | grep Taint
Taints:             <none>
```

From `--control-plan-node-count 2` but `--worker-node-count 0`:

```
$ kubectl --kubeconfig /home/smcginnis/.config/tanzu/tkg/unmanaged/moo/kube.conf get nodes
NAME                 STATUS     ROLES                  AGE    VERSION
moo-control-plane    NotReady   control-plane,master   108s   v1.21.5
moo-control-plane2   NotReady   control-plane,master   83s    v1.21.5
~
$ kubectl --kubeconfig /home/smcginnis/.config/tanzu/tkg/unmanaged/moo/kube.conf describe node moo-control-plane | grep Taint
Taints:             node-role.kubernetes.io/master:NoSchedule
```

**Comment by stmcginnis** [comment]
I can't find a way to change those taints via kind config. The note is pretty clear on behavior here:

https://github.com/kubernetes-sigs/kind/blob/e3864c0be8cc3b17380d7d98d7f2632212ef79b6/pkg/apis/config/v1alpha4/types.go#L154-L157

Perhaps we should start with only allowing `--worker-node-count` and leave `--control-plane-node-count` hard coded to 1 for now?

File: cli/cmd/plugin/unmanaged-cluster/cmd/create.go
Comment (jpmcb) [reply cli/cmd/plugin/unmanaged-cluster/cmd/create.go:67]: Good catch - these probably don't need shorthands and can just be `StringVar`

File: cli/cmd/plugin/unmanaged-cluster/config/config_test.go
Comment (jpmcb) [reply cli/cmd/plugin/unmanaged-cluster/config/config_test.go:222]: Guilty as charged 😂

**Comment by jpmcb** [comment]
> Perhaps we should start with only allowing --worker-node-count and leave --control-plane-node-count hard coded to 1 for now?

That was the behavior from https://github.com/vmware-tanzu/community-edition/pull/3139 where `--nodes` would denote the number of _total_ nodes deployed, including 1 control plane by default and the rest workers.

Maybe, the `kind` provider needs some extra logic where if
```
worker-nodes == 0 && control-plane nodes > 1:
  error "cannot deploy multiple control plane nodes with no workers. See [doc link] for more info"
```
?

I think it's still useful to expose the interface of setting control-plane nodes _and_ worker nodes, but the `kind` provider needs to handle the case where it can't deploy multiple CPs without a worker.

**Comment by stmcginnis** [comment]
Maybe we could get around this for now in our code with (pseudo):

```
if controlplanes > 1 and workernodes <1:
    error("multiple control plane nodes require at least one worker node for workload placement")
```

Otherwise, we do have the kubeconfig and we could get the list of nodes and iterate through the control plane nodes and untaint them. That's fairly simple, but could have a little more involved implementation.

**Comment by jpmcb** [comment]
+1 @stmcginnis  - very similar pseudo-code! 😂 

But yes, I agree, we could easily untaint them. Maybe something to tackle outside of this PR? If so, why don't we go with the error case you came up with and  can create a followup issue.

**Comment by stmcginnis** [comment]
> very similar

Oops, at first quick read I thought you were suggesting that was something `kind` itself should handle. So yeah, in total agreement - I think it makes sense for our `kind` provider to add that check for now. :D

**Comment by jpmcb** [comment]
Updated! 🚀

File: cli/cmd/plugin/unmanaged-cluster/cmd/create.go
Comment (joshrosso) [inline cli/cmd/plugin/unmanaged-cluster/cmd/create.go:67]: nit: I think we should remove `NOTE: Multiple control-plane nodes without a worker node is not supported`, namely because this is a decision for the provider implementation take make.

Comment (stmcginnis) [reply cli/cmd/plugin/unmanaged-cluster/cmd/create.go:67]: Maybe add "when using the 'kind' provider? It seems like we should have some mention of it if we are going to reject the request later on.

Or we could just set worker node count to 1 if not specified when control plane node count is more than 1.

Comment (jpmcb) [reply cli/cmd/plugin/unmanaged-cluster/cmd/create.go:67]: Very true - great catch Josh!

Comment (jpmcb) [reply cli/cmd/plugin/unmanaged-cluster/cmd/create.go]: Whoops - didn't refresh to see your comment Sean. Went ahead and removed it.

Setting the control-plane node to 1 regardless of the users input feels weird to me. But we should have something that communicates what's happening.

Here's an alternative approach I've considered as well:

```shell
# Create a cluster with 2 control plane nodes
# Installation will stale (not error out) and print warning message that user needs to remove taints manually
$ 

...

🚀 Creating cluster test
   Base image downloaded
   Cluster created
   To troubleshoot, use:

   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/test/kube.conf

⚠️ Warning! Kind based clusters with multiple control planes 
   and no worker nodes will NOT have workloads scheduled to them
   You must remove taints manually in order for installation to continue:

   kubectl --kubeconfig ~/.config/tanzu/tkg/unmanaged/test/kube.conf taint nodes --all node-role.kubernetes.io/master-
   
📧 Installing kapp-controller
   kapp-controller status: Pending...

----

# In a separate terminal, a user can remove the taints using the following command
$ ❯ kubectl --kubeconfig ~/.config/tanzu/tkg/unmanaged/test/kube.conf taint nodes --all node role.kubernetes.io/master-
node/test-control-plane untainted
node/test-control-plane2 untainted

-----

# Installation resumes now that kapp controller can be scheduled onto the control planes
📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

...
```

This workflow worked locally after removing the block that errors if there are no worker nodes to schedule to.

Comment (joshrosso) [reply cli/cmd/plugin/unmanaged-cluster/cmd/create.go]: I'm happy to revisit approaches outside this PR.

However, for now, I believe surfacing whether control plane / worker values are valid should be done in the cluster package, via providers. It's something providers can also surface in their pre-flight checks. I'm not too crazy about having conditionals in our flag help text.

**Review by joshrosso — APPROVED** [review]
works great :+1:

**Comment by joshrosso** [comment]
Let's wait for @stmcginnis to approve/request comments before merging.

2 is better than 1 :wink: .

**Review by stmcginnis — APPROVED** [review]
Good!

**Comment by jorgemoralespou** [comment]
@jpmcb has this been tested with multiple nodes exposing ports?

e.g. `tanzu unmanaged-cluster create life --control-plane-node-count 2 --worker-node-count 2 -p 80:80 -p 443:443`

**Comment by jpmcb** [comment]
@jorgemoralespou - 👀 good catch: this needed a new issue., created here: https://github.com/vmware-tanzu/community-edition/issues/3293

**Pull request merged**
